### PR TITLE
fix: dynamo batch write unprocessed

### PR DESCRIPTION
**Pull request by simone-sanfratello**

### CODE REVIEW

**Review by francardoso93 — CHANGES_REQUESTED** [review]

File: src/lib/storage.js [hunk 1]
```diff
@@ -68,12 +68,20 @@ async function createDynamoItem({ table, keyName, keyValue, data = {}, logger })
...
+      const response = await telemetry.trackDuration('dynamo-batch-inserts', sendCommand({ client: dynamoClient, command, logger }))
+      if (response?.UnprocessedItems && Object.keys(response.UnprocessedItems).length > 0) {
+        telemetry.increaseCount('dynamo-unprocessed-items')
+        logger.warn({ UnprocessedItems: JSON.stringify(response.UnprocessedItems) }, 'UnprocessedItems in BatchWriteItemCommand')
+        request = { RequestItems: response.UnprocessedItems }
+        await sleep(dynamoRetryDelay)
+        continue
+      }
+      done = true
+    } while (!done)
```
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply src/lib/storage.js:84]: Not sure it's the right thing to do, according to documentation, the unprocessed items should be processed again until they success

https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html

Which cases do you think need a max retries?

[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply src/lib/storage.js:84]: Also consider, the batch size is 25

**Review by francardoso93 — APPROVED** [review]

**Review by fabiolrodriguez — APPROVED** [review]

File: src/lib/storage.js [hunk 2]
```diff
@@ -68,12 +68,20 @@ async function createDynamoItem({ table, keyName, keyValue, data = {}, logger })
...
-      logger.warn({ UnprocessedItems: JSON.stringify(response.UnprocessedItems) }, 'UnprocessedItems in BatchWriteItemCommand')
-    }
+    let request = composeBatchInsert(batch)
+    let done
+    do {
+      const command = new BatchWriteItemCommand(request)
+      const response = await telemetry.trackDuration('dynamo-batch-inserts', sendCommand({ client: dynamoClient, command, logger }))
+      if (response?.UnprocessedItems && Object.keys(response.UnprocessedItems).length > 0) {
+        telemetry.increaseCount('dynamo-unprocessed-items')
+        logger.warn({ UnprocessedItems: JSON.stringify(response.UnprocessedItems) }, 'UnprocessedItems in BatchWriteItemCommand')
```
Comment (blucas) [inline src/lib/storage.js:78]: Question: Do we want to consider this a `warn` or an `info`?  If AWS treats this as a normal occurrence, then we may as well also.

File: src/lib/storage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (blucas) [reply src/lib/storage.js:84]: I think `maxRetries` may be of worth _if_ we believe that AWS may never successfully process all our items.  In that case, we'd be stuck in an endless loop.  My assumption is that all these items will successfully process.

**Review by blucas — APPROVED** [review]
Looks good.  Minor question though.

**Pull request merged**